File: 3.x.x/crates-utopia/user-agent/src/bot.rs
use serde::Serialize;

/// Detected bot / crawler metadata.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Bot {
    pub name: String,
    pub category: String,
}

impl Bot {
    /// Create a bot entry with the given name and category.
    pub fn new(name: &str, category: &str) -> Self {
        Self {
            name: name.to_string(),
            category: category.to_string(),
        }
    }

    /// Serialize to a flat map (PHP `toArray` shape).
    pub fn to_array(&self) -> BotArray {
        BotArray {
            name: self.name.clone(),
            category: self.category.clone(),
        }
    }
}

/// `Bot::to_array()` output.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct BotArray {
    pub name: String,
    pub category: String,
}

File: packages/user-agent/src/UserAgent/Bot.php
<?php

declare(strict_types=1);

namespace Utopia\UserAgent;

final readonly class Bot
{
    public function __construct(
        public string $name,
        public string $category = 'crawler',
    ) {}

    /**
     * @return array{name: string, category: string}
     */
    public function toArray(): array
    {
        return [
            'name' => $this->name,
            'category' => $this->category,
        ];
    }
}
